elastic-buffer-sim/src/Bittide/ClockControl/Strategies/Callisto.hs

```diff
@@ -2,81 +2,130 @@
+  -- | Speed change requested from clock multiplier
   Signal dom SpeedChange
-callisto clk rst ena ClockControlConfig{..} =
-  mealy clk rst ena go (initControlSt, 0)
+callisto targetCount updateEveryNCycles dataCounts =
+  mux shouldUpdate (D.toSignal b_kNext) (pure NoChange)
  where
-  go (ControlSt{..}, settleCounter) dataCounts
-    | settleCounter > cccSettlePeriod
-    = ((ControlSt x_kNext z_kNext b_kNext, 0), b_kNext)
-   where
-
-    -- see clock control algorithm simulation here:
-    -- https://github.com/bittide/Callisto.jl/blob/e47139fca128995e2e64b2be935ad588f6d4f9fb/demo/pulsecontrol.jl#L24
-    --
-    -- the constants here are chosen to match the above code.
-    k_p = 2e-4 :: Float
-    k_i = 1e-11 :: Float
-    r_k =
-      let
-        measuredSum = realToFrac (sum dataCounts)
-        targetCount = realToFrac (targetDataCount cccBufferSize)
-        nBuffers = realToFrac (length dataCounts)
-      in
-        measuredSum - targetCount * nBuffers
-    x_kNext =
-      x_k + p * r_k
-
-    c_des = k_p * r_k + k_i * realToFrac x_kNext
-    z_kNext = z_k + sgn b_k
-    fStep = 5e-4
-    c_est = fStep * realToFrac z_kNext
-    -- we are using 200kHz instead of 200MHz
-    -- (see https://github.com/clash-lang/clash-compiler/issues/2328)
-    -- so typical freq. is 0.0002 GHz
-    --
-    -- (this is adjusted by a factor of 100 because our clock corrections are
-    -- faster than those simulated in Callisto; we correct as often as hardware
-    -- allows)
-    p = 1e3 * typicalFreq where typicalFreq = 0.0002
-
-    b_kNext =
-      case compare c_des c_est of
-        LT -> SlowDown
-        GT -> SpeedUp
-        EQ -> NoChange
-
-    sgn NoChange = 0
-    sgn SpeedUp = 1
-    sgn SlowDown = -1
-
-  go (st, settleCounter) _ =
-    ((st, settleCounter + cccPessimisticPeriod), NoChange)
+  updateCounter = wrappingCounter updateEveryNCycles
+  shouldUpdate = updateCounter .==. 0
+  state = register initState (mux shouldUpdate updatedState state)
+  updatedState = D.toSignal $
+    ControlSt
+      <$> D.delayI (errorX "No start value [1]") x_kNext
+      <*> D.delayI (errorX "No start value [2]") z_kNext
+      <*> b_kNext
+
+  x_k :: DSignal dom 0 Float
+  x_k = D.fromSignal (_x_k <$> state)
+
+  z_k :: DSignal dom 0 (Signed 32)
+  z_k = D.fromSignal (_z_k <$> state)
+
+  b_k :: DSignal dom 0 SpeedChange
+  b_k = D.fromSignal (_b_k <$> state)
+
+  -- see clock control algorithm simulation here:
+  -- https://github.com/bittide/Callisto.jl/blob/e47139fca128995e2e64b2be935ad588f6d4f9fb/demo/pulsecontrol.jl#L24
+  --
+  -- the constants here are chosen to match the above code.
+  k_p, k_i, fStep :: forall m. DSignal dom m Float
+  k_p = pure 2e-4
+  k_i = pure 1e-11
+  fStep = pure 5e-4
+
+  r_k :: DSignal dom F.FromS32DefDelay Float
+  r_k = F.fromS32 $ D.fromSignal $
+    let
+      -- These bit coerces are fairly dangerous / unchecked.
+      --
+      -- TODO: Defend datatypes (why they don't overflow), or propagate errors up.
```

Comment:
Should this be done before merging?

Reply:
No, 32 bit should be plenty for our demos. I'm trying to speedrun getting things working on the board.